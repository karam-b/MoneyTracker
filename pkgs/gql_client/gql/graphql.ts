/* eslint-disable */
import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
  Date: { input: any; output: any; }
};

export type Category = {
  __typename?: 'Category';
  color?: Maybe<Scalars['String']['output']>;
  createdAt: Scalars['Date']['output'];
  createdBy: User;
  icon?: Maybe<Scalars['String']['output']>;
  id: Scalars['ID']['output'];
  title: Scalars['String']['output'];
  updatedAt: Scalars['Date']['output'];
};

export type CategoryInput = {
  color?: InputMaybe<Scalars['String']['input']>;
  icon?: InputMaybe<Scalars['String']['input']>;
  title: Scalars['String']['input'];
};

export type Entry = {
  __typename?: 'Entry';
  amount: Scalars['Float']['output'];
  category: Category;
  createdAt: Scalars['Date']['output'];
  createdBy: User;
  id: Scalars['ID']['output'];
  note?: Maybe<Scalars['String']['output']>;
  title: Scalars['String']['output'];
  updatedAt: Scalars['Date']['output'];
};

export type EntryInput = {
  amount?: InputMaybe<Scalars['Float']['input']>;
  category?: InputMaybe<Scalars['ID']['input']>;
  note?: InputMaybe<Scalars['String']['input']>;
  title: Scalars['String']['input'];
};

export type Mutation = {
  __typename?: 'Mutation';
  createManyCategories: Array<Category>;
  createManyEntries: Array<Entry>;
  createOneCategory: Category;
  createOneEntry: Entry;
  deleteManyCategories: Scalars['Int']['output'];
  deleteManyEntries: Scalars['Int']['output'];
  deleteOneCategory: Scalars['Boolean']['output'];
  deleteOneEntry: Scalars['Boolean']['output'];
  updateCurrentUser: User;
  updateOneCategory: Scalars['Boolean']['output'];
  updateOneEntry: Scalars['Boolean']['output'];
  updatePassword: Scalars['Boolean']['output'];
};


export type MutationCreateManyCategoriesArgs = {
  categories: Array<CategoryInput>;
};


export type MutationCreateManyEntriesArgs = {
  entries: Array<EntryInput>;
};


export type MutationCreateOneCategoryArgs = {
  category: CategoryInput;
};


export type MutationCreateOneEntryArgs = {
  entry: EntryInput;
};


export type MutationDeleteManyCategoriesArgs = {
  ids: Array<Scalars['ID']['input']>;
};


export type MutationDeleteManyEntriesArgs = {
  ids: Array<Scalars['ID']['input']>;
};


export type MutationDeleteOneCategoryArgs = {
  id: Scalars['ID']['input'];
};


export type MutationDeleteOneEntryArgs = {
  id: Scalars['ID']['input'];
};


export type MutationUpdateCurrentUserArgs = {
  user: UserInput;
};


export type MutationUpdateOneCategoryArgs = {
  category: CategoryInput;
  id: Scalars['ID']['input'];
};


export type MutationUpdateOneEntryArgs = {
  entry: EntryInput;
  id: Scalars['ID']['input'];
};


export type MutationUpdatePasswordArgs = {
  password: Scalars['String']['input'];
};

export type Query = {
  __typename?: 'Query';
  getAllCategories: Array<Category>;
  getAllEntries: Array<Entry>;
  getCurrentUser: User;
  getOneCategory?: Maybe<Category>;
  getOneEntry?: Maybe<Entry>;
};


export type QueryGetOneCategoryArgs = {
  id: Scalars['ID']['input'];
};


export type QueryGetOneEntryArgs = {
  id: Scalars['ID']['input'];
};

export type User = {
  __typename?: 'User';
  avatar?: Maybe<Scalars['String']['output']>;
  createdAt: Scalars['Date']['output'];
  displayName?: Maybe<Scalars['String']['output']>;
  email: Scalars['String']['output'];
  id: Scalars['ID']['output'];
  providers: Scalars['String']['output'];
  updatedAt: Scalars['Date']['output'];
};

export type UserInput = {
  avatar?: InputMaybe<Scalars['String']['input']>;
  displayName?: InputMaybe<Scalars['String']['input']>;
};

export type Get_Current_ProfileQueryVariables = Exact<{ [key: string]: never; }>;


export type Get_Current_ProfileQuery = { __typename?: 'Query', getCurrentUser: { __typename?: 'User', id: string, email: string, displayName?: string | null, avatar?: string | null, providers: string, createdAt: any, updatedAt: any } };


export const Get_Current_ProfileDocument = {"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"get_current_profile"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"getCurrentUser"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"Field","name":{"kind":"Name","value":"email"}},{"kind":"Field","name":{"kind":"Name","value":"displayName"}},{"kind":"Field","name":{"kind":"Name","value":"avatar"}},{"kind":"Field","name":{"kind":"Name","value":"providers"}},{"kind":"Field","name":{"kind":"Name","value":"createdAt"}},{"kind":"Field","name":{"kind":"Name","value":"updatedAt"}}]}}]}}]} as unknown as DocumentNode<Get_Current_ProfileQuery, Get_Current_ProfileQueryVariables>;