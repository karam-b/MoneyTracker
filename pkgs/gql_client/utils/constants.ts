export const GRAPHQL = "http://localhost:8812/graphql";
export const REST_API = "http://localhost:8812";
export const HELPER_SERVER = "http://localhost:8813";
