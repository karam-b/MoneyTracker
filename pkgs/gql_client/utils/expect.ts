import { expect } from "@jest/globals";
import { Response } from "node-fetch";

export function expectResponse(res: Response) {
  expect(String(res.status)[0]).toBe("2");
  expect(res.headers.get("x-token"));
}

export const profileRest = {
  created_at: expect.any(Number),
  updated_at: expect.any(Number),
  id: expect.any(String),
};

export const profile = {
  createdAt: expect.any(Number),
  updatedAt: expect.any(Number),
  id: expect.any(String),
};
