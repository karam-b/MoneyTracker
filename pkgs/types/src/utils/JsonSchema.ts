import { JSONSchema7 } from 'json-schema'
import refs from '../../dist/helpers/refs'

type Refed = Omit<JSONSchema7, '$ref'>

export type JSONSchema =
  | { $ref: refs }
  | (Omit<
      Refed,
      | 'items'
      | '$defs'
      | 'additionalItems'
      | 'contains'
      | 'properties'
      | 'patternProperties'
      | 'additionalProperties'
      | 'propertyNames'
      | 'if'
      | 'then'
      | 'else'
      | 'allOf'
      | 'anyOf'
      | 'oneOf'
      | 'not'
      | 'definitions'
    > & {
      items?: JSONSchema | JSONSchema[] | undefined
      $defs?:
        | {
            [key: string]: JSONSchema
          }
        | undefined
      additionalItems?: JSONSchema | undefined
      contains?: JSONSchema | undefined
      properties?: { [key: string]: JSONSchema } | undefined
      patternProperties?: { [key: string]: JSONSchema } | undefined
      additionalProperties?: JSONSchema
      propertyNames?: JSONSchema | undefined
      if?: JSONSchema | undefined
      then?: JSONSchema | undefined
      else?: JSONSchema | undefined
      allOf?: JSONSchema | undefined
      anyOf?: JSONSchema | undefined
      oneOf?: JSONSchema | undefined
      not?: JSONSchema | undefined
      definitions?: { [key: string]: JSONSchema } | undefined
    })
