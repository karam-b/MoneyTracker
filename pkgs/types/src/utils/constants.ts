export const BASE_URL = 'https://backend-json.pocket-app.karam.page'
