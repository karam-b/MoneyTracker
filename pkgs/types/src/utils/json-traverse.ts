import traverse from 'json-schema-traverse'
import resolve from './resolve'

export function getDefinitions(res: NonNullable<object>) {
  const refs = Array<string>()
  traverse(res, {
    cb: function (
      schema,
      JSONPointer,
      rootSchema,
      parentJSONPointer,
      parentKeyword,
    ) {
      if (parentKeyword === 'definitions') {
        // refs.push('#' + JSONPointer)
        if ('$id' in res && typeof res.$id === 'string') {
          refs.push(res.$id + '#' + JSONPointer)
          return
        }
        refs.push(+'#' + JSONPointer)
      }
    },
  })

  return refs
}

/**
 * test case #1 
const res = getDefinitions({
  $id: '/test',
  definitions: {
    test: {
      type: 'string',
    },
    boo: {
      type: 'object',
      properties: {
        test: {
          $ref: '#/definitions/test',
        },
      },
    },
  },
})
console.log({ res }) // { res: [ '/test#/definitions/test', '/test#/definitions/boo' ] }
// expect(res).toEqual(['/test#/definitions/test', '/test#/definitions/boo'])
*/

export function getIds(res: NonNullable<object>) {
  const refs = Array<string>()

  traverse(res, {
    cb: function (schema, JSONPointer) {
      if (JSONPointer === '') return

      if (!('$id' in schema && typeof schema.$id === 'string')) return

      if (!schema.$id) return

      const from = (res as { $id: string })?.$id || ''

      if (
        !(
          schema.$id.search(/^#[a-zA-Z.]/) === 0 ||
          schema.$id.search(/^\/[a-zA-Z.]/) === 0
        )
      ) {
        console.log('not tested', {
          to: schema.$id,
          from,
          res: resolve(from, schema.$id),
        })
      }

      refs.push(resolve(from, schema.$id))
    },
  })

  return refs
}

/**
 * test case #2

const res2 = getIds({
  $id: 'http://test.com/test',
  type: 'object',
  properties: {
    test: {
      $id: '#test',
      type: 'string',
    },
    boo: {
      $id: '#boo',
      type: 'object',
      properties: {
        test: {
          $id: '#nested',
          type: 'string',
        },
        absolute: {
          $id: '/absolute',
          type: 'string',
        },
      },
    },
  },
})

console.log({ res2 }) // { res: [ 'http://test.com/test#test', 'http://test.com/test#boo', 'http://test.com/test#nested', 'http://test.com/absolute' ] }
 */
