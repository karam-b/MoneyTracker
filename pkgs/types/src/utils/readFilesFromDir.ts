/* eslint-env node */
/// <reference types="node" />
import fs from 'fs'
import path from 'path'

export function FilesFromDir(dir: string): string[] {
  const files = Array<string>()
  const files_or_dirs = fs.readdirSync(dir, {
    withFileTypes: true,
  })

  files_or_dirs.forEach(file_or_dir => {
    if (file_or_dir.isDirectory()) {
      const recursive = FilesFromDir(`${dir}/${file_or_dir.name}`)
      files.push(...recursive)
      return
    }

    files.push(path.join(dir, file_or_dir.name))
  })

  return files
}

export default function readFilesFromDir<T = string>(
  dir: string,
  transform?: (input: { out: string; name: string }) => T,
) {
  const files = FilesFromDir(dir)

  if (transform)
    return files
      .map(f => [f, fs.readFileSync(f, 'utf-8')])
      .map(([name, out]) => transform({ name, out })) as T[]

  return files.map(f => fs.readFileSync(f, 'utf-8')) as T[]
}
