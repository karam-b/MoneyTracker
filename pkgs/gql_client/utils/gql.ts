import { graphql } from "../gql/gql";

export default function gql(literals, ...placeholders) {
  // const input = literals.join("");
  const input = literals.reduce((acc, cur, idx) => {
    return acc + cur + (placeholders[idx] || "");
  }, "");

  const rt = graphql(input);

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  rt.__my_input = input;
  return graphql(input);
}

export function getGqlInput(rt: ReturnType<typeof gql>) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return rt.__my_input;
}
