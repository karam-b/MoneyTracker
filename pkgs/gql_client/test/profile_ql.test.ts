import { afterAll, beforeAll, expect, it } from "@jest/globals";
import gql, { getGqlInput } from "../utils/gql";
import fetch from "node-fetch";
import { GRAPHQL, HELPER_SERVER, REST_API } from "../utils/constants";
import { request } from "graphql-request";
import { profile } from "../utils/expect";

const ctx: any = {
  ids: [],
};

/* Graphql */
const get_current_user = gql`
  query get_current_profile {
    getCurrentUser {
      id
      email
      displayName
      avatar
      providers
      createdAt
      updatedAt
    }
  }
`;

beforeAll(async () => {
  const res = await fetch(REST_API + "/auth/local/register", {
    method: "POST",
    headers: {
      content: "application/json",
      authorization:
        "Basic " + Buffer.from("testing_profile@m.c:pass").toString("base64"),
    },
    body: JSON.stringify({}),
  });
  const json: any = await res.json();
  ctx.user = Object.assign({}, json);
  ctx.ids.push(json.id);
  ctx.user.token = res.headers.get("x-token");
  ctx.user.headers = {
    authorization: "Bearer " + ctx.user.token,
  };
});

it("wtf", async () => {
  const data = await request(GRAPHQL, get_current_user, {}, ctx.user.headers);

  // expect(data.getCurrentUser).toMatchSnapshot(profile);
  expect(data.getCurrentUser).toMatchSnapshot(profile);
});

afterAll(async () => {
  await Promise.all(
    ctx.ids.map((id: string) =>
      fetch(HELPER_SERVER + "/remove_user_data/" + id)
    )
  );
});
