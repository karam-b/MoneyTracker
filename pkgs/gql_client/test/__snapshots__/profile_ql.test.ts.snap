// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wtf 1`] = `
Object {
  "avatar": null,
  "createdAt": "2023-08-26T05:19:18.300734+00:00",
  "displayName": null,
  "email": "testing_profile@m.c",
  "id": "108",
  "providers": "local",
  "updatedAt": "2023-08-26T05:19:18.300734+00:00",
}
`;
